Core/src/com/infiniteautomation/mango/spring/service/ConfigurationTemplateService.java

```diff
@@ -0,0 +1,155 @@
+     * @param filePath
+     * @return
+     * @throws IOException
+     * @throws PermissionException
+     */
+    public List<Map<String, Object>> generateConfig(String fileStore, String filePath) throws IOException,
+            PermissionException {
+        permissionService.ensureAdminRole(Common.getUser());
+
+        Path path = fileStoreService.getPathForRead(fileStore, filePath);
+        try (Reader in = new InputStreamReader(Files.newInputStream(path), Charset.defaultCharset())) {
+            CSVReader csvReader = new CSVReader(in);
+            List<Map<String, Object>> readFileStructure = readCSV(csvReader);
+            List<Map<String, String>> keys = new ArrayList<>(){{
+                add(new HashMap<>(){{
+                    put("groupKey", "site");
+                    put("childrenKey", "dataHalls");}});
+                add(new HashMap<>(){{
+                    put("groupKey", "dataHall");
+                    put("childrenKey", "devices");}});
+            }};
+
+            List<Map<String, Object>> result = groupByKeys(readFileStructure, keys);
+            return result;
+        }
+    }
+    public String generateTemplate(String fileStore, String filePath, String template) throws IOException,
+            PermissionException {
+        MustacheFactory mf = new DefaultMustacheFactory();
+        List<Map<String, Object>> sites = generateConfig(fileStore, filePath);
+
+        ClassPathResource cp = new ClassPathResource(template);
+        Mustache m = mf.compile(cp.getPath());
+        StringWriter writer = new StringWriter();
+        m.execute(writer, sites).flush();
+        System.out.println("Testing result" + writer.toString());
+        return writer.toString();
+    }
+
+    public static List<Map<String, Object>> groupByKey(List<Map<String, Object>> array, Map<String, String> key) {
+        Map<Object, Map<String, Object>> map = new HashMap<>();
+        for (Map<String, Object> item : array) {
+            Object keyValue = item.get(key.get("groupKey"));
+            if (!map.containsKey(keyValue)) {
+                Map<String, Object> group = new HashMap<>();
+                group.put(key.get("groupKey"), keyValue);
+                group.put(key.get("childrenKey"), new ArrayList<Map<String, Object>>());
+                map.put(keyValue, group);
+            }
+            ((List)map.get(keyValue).get(key.get("childrenKey"))).add(item);
+        }
+        List<Map<String, Object>> groups = new ArrayList<>(map.values());
+
+        // find common properties from the children and copy them to the group
+        for (Map<String, Object> group : groups) {
+            Map<String, Object> common = findCommonProperties((List<Map<String, Object>>) group.get(key.get("childrenKey")));
+            group.putAll(common);
+        }
+
+        return groups;
+    }
+
+    public static List<Map<String, Object>> groupByKeys(List<Map<String, Object>> data, List<Map<String, String>> keys) {
```

Comment:
Any specific reason these are public and static?

Reply:
none, changed to private methods.